### PR TITLE
Add reduced train policy metrics

### PR DESCRIPTION
**Pull request by Hyperion-shuo**
## Summary

Add reduced train policy metrics to monitor the mismatch between current policy logprobs and rollout-time old logprobs.

The new metrics include:

- policy ratio min / max
- policy ratio absolute deviation mean
- approximate KL estimates `kl1` and `kl3`
- low / high clip fractions

The rollout-time old logprobs are generated on GPU during rollout. These metrics are therefore used to monitor policy lag between rollout-time sampling/logprob computation and subsequent training, which is different from train-inference implementation mismatch metrics.

## Motivation

Policy lag health is an important reference when tuning `RLDisaggregatedTrainer`. In disaggregated RL training, rollout / training resource allocation and related hyperparameters can change how stale rollout samples are by the time they are optimized.

These metrics help decide whether the current train / rollout resource ratio and update schedule are reasonable.

## Example

image

### CODE REVIEW

**Comment by hhaAndroid** [comment]
@claude review

**Comment by hhaAndroid** [comment]
这些新加入的训练指标，会在训练时候每一次都随着 step 打印吗？那终端的输出是不是非常多？麻烦把终端打印输出贴一下，看看效果

**Comment by Hyperion-shuo** [comment]
@hhaAndroid 
image

这些 train policy metrics 已经做过 reduce，现在每个 rank 每个 train step 都打印到终端，确实带来冗余日志。

我在每个 rank 的 worker INFO 终端日志里过滤掉这批新增的 train_policy metrics。

File: xtuner/v1/rl/loss/grpo_loss.py [hunk 1]
```diff
@@ -166,9 +166,33 @@ def loss_fn(
...
-        ratio = ratio * (shifted_labels != self.loss_cfg.ignore_idx).float()
-        extra_info = {"max_ratio": ratio.max()}
+        valid_mask = shifted_labels != self.loss_cfg.ignore_idx
+        valid_float = valid_mask.float()
+        cliprange_low = self.loss_cfg.policy_loss_cfg.get("cliprange_low")
+        cliprange_high = self.loss_cfg.policy_loss_cfg.get("cliprange_high")
+
+        log_ratio = logprobs.detach() - old_logprobs.detach()
+        log_ratio_safe = torch.clamp(log_ratio, min=-20.0, max=20.0)
+        ratio = torch.exp(log_ratio_safe)
```
[the remark this answers is not in the document]
Comment (Hyperion-shuo) [reply xtuner/v1/rl/loss/grpo_loss.py:176]: `kl1` is linear in `log_ratio`, so it does not need clamping for numerical stability. Keeping it on the raw `log_ratio` preserves extreme policy drift signals, which is useful for diagnosing policy lag.

`kl3` needs `ratio = exp(log_ratio)`, so it uses the clamped `log_ratio_safe` to avoid overflow. This means `kl3` is the numerically stable approximate KL metric, while `kl1` keeps the raw signed drift signal.

So I prefer to keep the current behavior.

File: xtuner/v1/rl/loss/base_loss.py [hunk 2]
```diff
@@ -249,3 +250,58 @@ def compute_rollout_is(
...
+            }
+        )
+    for output_key, sum_key in output_keys.items():
+        extra_info_dict[output_key] = train_policy_values[sum_key] / valid_count if valid_count > 0 else 0.0
+
+    ratio_max = train_policy_values["reduced_train_policy_ratio_max"]
+    ratio_min = train_policy_values["reduced_train_policy_ratio_min"] if valid_count > 0 else 0.0
+    extra_info_dict["reduced_train_policy_ratio_max"] = ratio_max
+    extra_info_dict["reduced_train_policy_ratio_min"] = ratio_min
+    # legacy metric，keep here
```
[the remark this answers is not in the document]
Comment (Hyperion-shuo) [reply xtuner/v1/rl/loss/base_loss.py:305]: The legacy `max_ratio` is kept for backward compatibility, I will note this semantic change in the PR description.

**Comment by hhaAndroid** [comment]
@Hyperion-shuo 对的，不需要所有指标都打印到终端，特别是 train_policy  包括 min max std 等等对同一个指标的多个维度打印信息，但是可以保存到 rank log 里面。

**Comment by Hyperion-shuo** [comment]
image

image


train_step的日志已经保持和之前一致，tensorboard里reduced_train_policy指标正常显示。

File: xtuner/v1/rl/loss/base_loss.py [hunk 3]
```diff
@@ -249,3 +250,58 @@ def compute_rollout_is(
...
+        "reduced_train_policy_kl3": "reduced_train_policy_kl3_sum",
+    }
+    if has_clip_count:
+        output_keys.update(
+            {
+                "reduced_train_policy_clip_frac_low": "reduced_train_policy_clip_low_count",
+                "reduced_train_policy_clip_frac_high": "reduced_train_policy_clip_high_count",
+            }
+        )
+    for output_key, sum_key in output_keys.items():
```
[the remark this answers is not in the document]
Comment (Hyperion-shuo) [reply xtuner/v1/rl/loss/base_loss.py:298]: fix in new commit

File: xtuner/v1/rl/loss/grpo_loss.py [hunk 4]
```diff
@@ -166,9 +166,33 @@ def loss_fn(
...
 
         assert old_logprobs is not None
-        ratio = (logprobs - old_logprobs.detach()).exp()
-        ratio = ratio * (shifted_labels != self.loss_cfg.ignore_idx).float()
-        extra_info = {"max_ratio": ratio.max()}
+        valid_mask = shifted_labels != self.loss_cfg.ignore_idx
+        valid_float = valid_mask.float()
+        cliprange_low = self.loss_cfg.policy_loss_cfg.get("cliprange_low")
+        cliprange_high = self.loss_cfg.policy_loss_cfg.get("cliprange_high")
+
```
[the remark this answers is not in the document]
Comment (Hyperion-shuo) [reply xtuner/v1/rl/loss/grpo_loss.py:173]: This is intentional. The policy loss path is unchanged and still uses the original ratio with gradients. The new ratio is metric-only, detached.

**Comment by hhaAndroid** [comment]
@claude review

**Review by windreamer — APPROVED** [review]
